src/Sylius/Component/Core/Promotion/Action/UnitDiscountPromotionActionCommand.php

```diff
@@ -67,7 +67,21 @@ protected function removeUnitOrderItemAdjustments(OrderItemUnitInterface $unit,
     protected function addAdjustmentToUnit(OrderItemUnitInterface $unit, int $amount, PromotionInterface $promotion): void
     {
         $adjustment = $this->createAdjustment($promotion, AdjustmentInterface::ORDER_UNIT_PROMOTION_ADJUSTMENT);
-        $adjustment->setAmount(-$amount);
+
+        /** @var OrderItemInterface $orderItem */
+        $orderItem = $unit->getOrderItem();
+
+        /** @var ProductVariantInterface $variant */
+        $variant = $orderItem->getVariant();
+
+        /** @var OrderInterface $order */
+        $order = $orderItem->getOrder();
+
+        $channel = $order->getChannel();
+
+        $minimumPrice = $variant->getChannelPricingForChannel($channel)->getMinimumPrice();
+
+        $adjustment->setAmount($this->calculate($unit->getTotal(), $minimumPrice, -$amount));
```

Comment:
I'm wondering if this should not be extracted logic - assigning/calculating amount of x or variants minimal price